### ISSUE TITLE
[RFE] Don't automatically uninstall apps declared via `catalog.cattle.io/auto-install`

### ISSUE THREAD

**Issue by viccuad** [rancher/terraform-provider-rancher2#1195]
### Is your feature request related to a problem? Please describe.

Currently, on uninstall of app `A`, terraform automatically uninstalls also whatever app was declared in app `A` via `catalog.cattle.io/auto-install` ([this code](https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_app_v2.go#L184-L196)).

This behaviour is particular to the terraform provider, and is not followed by Rancher UI.

IMHO this behaviour is dangerous:
- You can have 2 separate charts both declare the same `catalog.cattle.io/auto-install: my-crds=1.0.0`. On uninstall of any of those chart releases, `my-crds` chart gets removed, breaking the remaining chart release in the system as it is now missing the needed CRDs. This is the case for `kubewarden-controller` and `kubewarden-defaults`, as both depend on `kubewarden-crds`. See issue raised by [community here](https://github.com/kubewarden/helm-charts/issues/272).
- You can have several charts declare `catalog.cattle.io/auto-install: rancher-monitoring`. On uninstall of one of the chart release, the remaining charts are now broken.

### Describe the solution you'd like

Make the terraform behaviour on `catalog.cattle.io/auto-install` optional and skippable on uninstall of chart `A`.

Maybe it it could be better improved if it would automatically skip a chart `B` specificed via `catalog.cattle.io/auto-install` , if that chart `B` has a `provides-gvr` for some CRDs that the chart `A` has in a `requires-gvr`. But this maybe doesn't fully solve transitive dependencies, and charts that provide or depend on several gvrs, as one cannot list several on those annotations. But this needs some consideration.

For the current behaviour (on `catalog.cattle.io/auto-install`)  in Terraform to work, it would be needed to have a proper dependency management system for Helm charts and/or OCI artifacts. This dependency system would know about dependencies, optional dependencies, their versions, if they are namespaced, and to delegate to it if a chart release is not needed in the system anymore or can be scheduled for removal.

### Describe alternatives you've considered

- Forking charts, so one can overwrite the `catalog.cattle.io/auto-install` annotations, effectively disabling the stated dependency and managing those dependencies manually.

- Forking the terraform provider and commenting [this code](https://github.com/rancher/terraform-provider-rancher2/blob/master/rancher2/resource_rancher2_app_v2.go#L184-L196).

- Adding a dependency system that is aware of rancher annotations and can be used as a drop-in substitute of Helm binary and library: https://hypper.io (disclaimer, helped write that thing).

### Additional context

See https://github.com/kubewarden/helm-charts/issues/272

**Comment by mpepping** [comment]
Since this issue is Rancher specific, triggered by Rancher specific annotations, I agree that it is best to use the same logic as the Rancher ui/api uses: `catalog.cattle.io/auto-install` values should be used for installation tasks only. 

The auto-install annotation mainly aims on installing CRD-charts, as that is [not well covered by Helm](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations). Automatically uninstalling CRD-charts impact other installed charts that may depend on that same CRD-chart.

Maybe at minimum, an `if` or toggle should be in place that allow to skip parsing the `catalog.cattle.io/auto-install` annotation for specific charts managed by a [rancher2_app_v2](https://registry.terraform.io/providers/rancher/rancher2/latest/docs/resources/app_v2) resource.

**Comment by thsnielsen** [comment]
@viccuad would you please add the jira label (it has the same name)

**Comment by viccuad** [comment]
@thsnielsen I don't have a Jira ticket for this, yet the customer raised hit this bug at that time, and it keeps being a maintenance issue.

I still maintain that automatically uninstalling (CRD) charts is dangerous.

The customer is currently forced to use the 3rd alternative I provided on the post: fork Helm charts and remove the `catalog.cattle.io/auto-install` annotation manually.

Please get in contact if I can help move this further.

**Comment by viccuad** [comment]
This is SURE-7863 (I don't have perms to add labels).

**Comment by matttrach** [comment]
I agree with @mpepping, the Terraform provider shouldn't alter the user experience provided by the API. Terraform providers should be an interface/adapter/SDK for the API they connect to. Unfortunately, this code has been in place for 4 years. Changing this behavior means altering the experience of current users which is a breaking change. As such the change will need to be in the next major release (v7 of the provider) (which alters the version compatibility from Rancher 2.8 to Rancher 2.11)

**Comment by matttrach** [comment]
PR to master - https://github.com/rancher/terraform-provider-rancher2/pull/1467 correlating to v7, Rancher v2.11

**Comment by matttrach** [comment]
backport to v6  - https://github.com/rancher/terraform-provider-rancher2/pull/1472 correlating to Rancher v2.10
